flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/LegacyTypeInfoDataTypeConverter.java

```diff
@@ -216,6 +216,18 @@ else if (canConvertToTimestampTypeInfoLenient(dataType)) {
 			return Types.SQL_TIMESTAMP;
 		}
 
+		// relax the precision constraint as LocalDateTime can store the highest precision
+		else if (hasRoot(logicalType, LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE) &&
```

Comment:
Maybe you can take a look to above, there are some comments and limitation for `java.sql.Timestamp`.
And others... If you want to modify, maybe you should modify all or just respect their rules. (Like in `canConvertToTimestampTypeInfoLenient`).